[Go SDK] Spanner Batched Reads

christogav opened the pull request:
Proposal to add batched reads to Go spannerio. Code fairy also refactored some of the spannerio package for improved comprehension and maintenance efforts.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/workflows/Build%20python%20source%20distribution%20and%20wheels/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/workflows/Python%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/workflows/Java%20Tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[![Go tests](https://github.com/apache/beam/workflows/Go%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

lostluck commented:
Tagging myself in to review this tomorrow.

With the Partitions, we can likely make this a proper SDF (or build that in) so things can scale better. But I'd understand if the preference is to get this into next week's 1.47 cut.

christogav commented:
@lostluck I will have a go at SDF following the textio example tonight.

lostluck commented on sdks/go/pkg/beam/io/spannerio/generate_partitions_test.go:
```diff
@@ -0,0 +1,57 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package spannerio provides an API for reading and writing resources to
+// Google Spanner datastores.
+
+package spannerio
```
```suggestion
package spannerio
```

lostluck commented on sdks/go/pkg/beam/io/spannerio/read_batch.go:
```diff
@@ -0,0 +1,140 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package spannerio provides an API for reading and writing resources to
+// Google Spanner datastores.
+package spannerio
```
```suggestion

package spannerio
```

lostluck commented on sdks/go/pkg/beam/io/spannerio/write.go:
```diff
@@ -0,0 +1,124 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package spannerio provides an API for reading and writing resouces to
+// Google Spanner datastores.
+package spannerio
```
```suggestion

package spannerio
```

lostluck commented on sdks/go/pkg/beam/io/spannerio/test_utils.go:
```diff
@@ -0,0 +1,67 @@
+	}
+}
+
+func createFakeClient(address string, database string) (*spanner.Client, *db.DatabaseAdminClient, func(), error) {
```
We can simplify things for the testing since Spanner has an env variable override we can use.

See `SPANNER_EMULATOR_HOST` in the spannertest docs.
https://pkg.go.dev/cloud.google.com/go/spanner/spannertest#section-readme

This lets us use the normal client paths, and may be able to work better on portable runners at least for unit test purposes.

Alternatively, we add another exported field to the spannerFn to put in the server address, and when that's present, we provide the appropriate connection option for the fake override. While this does move some test specific code into the production code, it makes testing much easier.

christogav replied:
I've stuck with this approach for admin clients (due to no support for the env var) and in unit tests to support parallelism (is that really a concern?).

For the integration test preferred SPANNER_EMULATOR_HOST.

Are we ok with this approach?

lostluck commented on sdks/go/pkg/beam/io/spannerio/read_batch_integration_test.go:
```diff
@@ -0,0 +1,126 @@
+
+	if _, err := iter.Next(); err == nil {
+		return nil
+	}
```
This section doesn't appear to be doing anything WRT "populating Spanner". If it's necessary, please explain why in a comment, and augment the returned error with this sort of detail as well. Otherwise, please remove.

lostluck commented on sdks/go/pkg/beam/io/spannerio/read_batch_integration_test.go:
```diff
@@ -0,0 +1,126 @@
+	passert.Count(s, rows, "Should have 4 rows", 4)
+}
+
+func populateSpanner(ctx context.Context, admin *database.DatabaseAdminClient, db string, client *spanner.Client) error {
```
Since this is part of test set up it should take in a `*testing.T` and call `t.Helper()`. All errors should then be logged with the `t.Error` or `t.Fatal` etc, assuming they block execution of the test. 

As it stands, all the errors are ignored, which would make debugging difficult.

See https://google.github.io/styleguide/go/decisions#test-helpers

lostluck commented on sdks/go/pkg/beam/io/spannerio/read_batch_integration_test.go:
```diff
@@ -0,0 +1,126 @@
+		return err
+	}
+
+	if err := op.Wait(context.Background()); err != nil {
```
A context is being passed in, use that instead of creating a new one or comment why the passed in context is being ignored.

lostluck commented on sdks/go/pkg/beam/io/spannerio/read_batch_integration_test.go:
```diff
@@ -0,0 +1,126 @@
+)
+
+func TestExampleQueryBatch(t *testing.T) {
+	if !*integrationTests {
+		t.Skip("Not running in integration test mode.")
+	}
```
Please put integration tests in an appropriate directory in https://github.com/apache/beam/tree/master/sdks/go/test/integration/io and use the existing infrastructure for integration tests. This will allow the test to run on several runners.

Although, I'll note that as written, this test won't work as it's using a fake client/server, like the other tests, but notably, the [spannertest fake doesn't support partitions at this time](https://pkg.go.dev/cloud.google.com/go/spanner/spannertest#section-readme)

It would be important for any integration tests against real spanner would need to take care to clean up after itself so we don't waste beam project resources.

christogav replied:
I've moved this to the integration test directory and corrected the issue of using the fake client/server.

lostluck reviewed (changes requested):
Style comments aside, is there any reason we shouldn't just make the standard Read and Query calls use an SDF with partitioned queries under the hood? Adding splitting at that level improves existing pipelines when they update Beam, and avoids unnecessary thinking about "should I ReadBatch, or Read?". 

In particular when without an SDF the benefits of ReadBatch here aren't clear.

lostluck reviewed (commented):
Also, please avoid mixing such a large refactoring for maintenance/clarity with a functional change.

Pick one or the other to do first. Mixing them makes it harder to review, and I'm forced to treat all changed code as new, rather than being able to treat much of it as existing code that is just moved around.

christogav commented on sdks/go/pkg/beam/io/spannerio/generate_partitions.go:
```diff
@@ -0,0 +1,149 @@
+
+	if txn == nil {
+		panic("unable to create a spanner transaction")
+	}
```
Removed.

christogav commented:
> Style comments aside, is there any reason we shouldn't just make the standard Read and Query calls use an SDF with partitioned queries under the hood? Adding splitting at that level improves existing pipelines when they update Beam, and avoids unnecessary thinking about "should I ReadBatch, or Read?".
> 
> In particular when without an SDF the benefits of ReadBatch here aren't clear.

@lostluck Not all queries can be partitioned and the value of forcing partitioned reading on an implementor in lower volume scenarios may(?) decrease performance, not increase it. I'm not sure that implementors can avoid consideration of whether batched reads are appropriate. I believe the java maintainers must have faced a similar dilemma as they have Naive/Batch also.

Thoughts?

lostluck commented:
> 

Thanks for your patience, my weeks are presently busier than desired.

My inclination is that the DoFn should do the best with what it's given, and avoid the user making unnecessary choices. We shouldn't force users to make a choice, when the choice is an efficient pipeline or not, and we can programatically figure that out for them.

We might be getting crossed between "queries" and "reads". Per https://cloud.google.com/spanner/docs/reads#read_data_in_parallel `You can perform any read API operation in parallel using the Spanner client libraries.` Right now we implement both the same DoFn (queryFn) but there's no reason that non-query Reads, which are parallelizable by nature, doesn't just always use a batch path.

If the query is unable to be partitioned, then the user isn't able to improve performance themselves anyways. 

If spanner only returns a single partition to a "partition query" request, then we have our answer, and the user can't do any better themselves. So really, the question is what's the consequence to asking spanner to partition an unpartionable query? If we get no error, and a single partition back, I say we stick to a single implementation.

If not, then we can still always do batches for reads, and leave queries for now. It's just how the previous contributor implemented things that they currently share an implementation. We can change that.

Heck, per the [same spanner doc]( https://cloud.google.com/spanner/docs/reads#read_data_in_parallel) we can always request the query plan first, at pipeline construction time, and see if the first thing is a [Distributed Union](https://cloud.google.com/spanner/docs/query-execution-operators#distributed-union), and if not, use a simpler single plan.

lostluck commented:
Whoops, got distracted by Timers, and I don't see any reply comments, so it was easier to miss this was ready for another look. Thanks for your patience. 

It's great to see this passing the presubmits, and thanks for all the fixes.

lostluck commented on sdks/go/pkg/beam/io/spannerio/query_options.go:
```diff
@@ -0,0 +1,76 @@
+
+// WithBatching sets whether we will use a batched reader. Batching is set to true by default, disable it when the
+// underlying query is not root-partitionable.
+func WithBatching(batching bool) func(opts *queryOptions) error {
```
I recommend adding a real type for the functional options, if only for documentation purposes:

`type QueryOptionFn func(*queryOptions)` and use that instead of the raw type. This makes the options show up nicely in documentation.  Eg. See how the options appear for `fileio.ReadOptionFn` and `fileio.MatchOptionFn` https://pkg.go.dev/github.com/apache/beam/sdks/v2@v2.47.0/go/pkg/beam/io/fileio

lostluck commented on sdks/go/pkg/beam/io/spannerio/read_batch.go:
```diff
@@ -0,0 +1,168 @@
+}
+
+// CreateInitialRestriction creates an offset range restriction representing
+// the partition's size in bytes.
```
```suggestion
// the number of rows in the partition..
```

lostluck commented on sdks/go/pkg/beam/io/spannerio/read_batch.go:
```diff
@@ -0,0 +1,168 @@
+// SplitRestriction splits each file restriction into blocks of a predetermined
+// size, with some checks to avoid having small remainders.
+func (f *readBatchFn) SplitRestriction(_ PartitionedRead, rest offsetrange.Restriction) []offsetrange.Restriction {
+	splits := rest.SizedSplits(blockSize)
```
Note that this approach also may have the same issue as textIO does with a fixed size split, see #25892. However, since Spanner has already done a bunch of splits for us, it might be moot.  Not required to fix this now, but just a note.

lostluck commented on sdks/go/pkg/beam/io/spannerio/read_batch.go:
```diff
@@ -0,0 +1,168 @@
+	return nil
+}
+
+// PartitionedRead holds relevant partition information to support partitioned reading from Spanner.
+type PartitionedRead struct {
```
We probably don't need to export this type. The fields need to be exported, but the struct itself does not.

lostluck commented:
Run Go PostCommit

lostluck reviewed (changes requested):
A couple of cleanup comments, but overall I'm very happy where this has ended up! Just a small handful of changes,.

I'm pretty sure the integration test will fail outside of the direct runner though since it's using a fake Spanner endpoint instead of the real spanner endpoint. I'm less certain about the other runners though, but it's ultimately filtered out. 

Dataflow won't be able to talk to the test container endpoint, and Environment Variables aren't automatically sent around. We don't have a generalized "copy and propagate these environment variable strings, and set them on the worker" option.

FYI the 2.48.0 cut is Wednesday May 17th. Please @ me to ensure I look at it once the last changes have been made.

lostluck commented:
waiting on author

christogav commented:
@lostluck Thank you so much for the reviews and feedback. I've addressed all feedback.

I've been pondering `ProcessElement` in read_batch.go and the performance impact of re-reading rows across splits. Are there any other different split strategies that might work better or do you think the current approach is ok?

lostluck commented:
Run Go PostCommit

lostluck commented:
> I've been pondering `ProcessElement` in read_batch.go and the performance impact of re-reading rows across splits. Are there any other different split strategies that might work better or do you think the current approach is ok?

So, this is a complicated question. As a rule, you're right. Re-reading and discarding repeatedly is likely to be a source of inefficiency. But that would require the runner in question to decide that the elements need sub splitting in the first place. If there's sufficient parallelism, it might not happen.

A runner that doesn't do dynamic splitting  might decide to eagerly have each eleemnt split, guaranteeing the bad behavior. A runner that can do dynamic splitting, might never split at that granularity, and never need to. BUT, if one of the rows is taking much longer to process, then it might want to have a given element split even at the cost of the re-read of rows.

So, arguably the way to avoid it is we move getting the batch partitions into CreateInitialRestriction, which complicates the restriction into containing both a offset range, and a slice of the transactions. Then the "element" that's getting split is just the query directly. The splitRestriction can then make smaller version that contain only a single partition each, instead of the offset. And finally, if a split request happens at *that* level, we can then move to using offset ranges and similar against the partitioned queries.

This avoids the rereading problem by default by only having initial splits be equal to the natural spanner partitioning, and only allows rereading when a sub element split needs to happen.

But TBH that's defintely something worth doing in another PR, where we can benchmark this solution, and then be able to compare it against any new approach first.

-------------

At this point, I'm going to be merging this in once the PostCommit passes (which is the Dataflow test set). If the current run fails, I'll filter out the test accordingly first.

lostluck commented:
> I'm pretty sure the integration test will fail outside of the direct runner though since it's using a fake Spanner endpoint instead of the real spanner endpoint. I'm less certain about the other runners though, but it's ultimately filtered out.
> 
> Dataflow won't be able to talk to the test container endpoint, and Environment Variables aren't automatically sent around. We don't have a generalized "copy and propagate these environment variable strings, and set them on the worker" option.

What I mentioned earlier to just surface it.

lostluck commented:
Oh didn't expect that: It didn't like sourcing the container from docker.io. No creds.

```
=== RUN   TestSpannerIO_QueryBatch
Testcontainers properties file has been found: /home/jenkins/.testcontainers.properties
2023/05/22 01:56:54 github.com/testcontainers/testcontainers-go - Connected to docker: 
  Server Version: 20.10.23
  API Version: 1.41
  Operating System: Ubuntu 20.04.3 LTS
  Total Memory: 104630 MB
2023/05/22 01:56:54 Failed to get image auth for docker.io. Setting empty credentials for the image: docker.io/testcontainers/ryuk:0.4.0. Error is:credentials not found in native keychain
2023/05/22 01:56:56 🐳 Creating container for image docker.io/testcontainers/ryuk:0.4.0
2023/05/22 01:57:08 ✅ Container created: 095e19f14ae9
2023/05/22 01:57:08 🐳 Starting container: 095e19f14ae9
2023/05/22 01:57:09 🚧 Waiting for container id 095e19f14ae9 image: docker.io/testcontainers/ryuk:0.4.0
2023/05/22 01:57:09 ✅ Container started: 095e19f14ae9
2023/05/22 01:57:09 Failed to get image auth for https://index.docker.io/v1/. Setting empty credentials for the image: cloud-spanner-emulator/emulator:latest. Error is:credentials not found in native keychain
2023/05/22 01:57:09 Failed to get image auth for https://index.docker.io/v1/. Setting empty credentials for the image: cloud-spanner-emulator/emulator:latest. Error is:credentials not found in native keychain
2023/05/22 01:57:11 Failed to get image auth for https://index.docker.io/v1/. Setting empty credentials for the image: cloud-spanner-emulator/emulator:latest. Error is:credentials not found in native keychain
2023/05/22 01:57:15 Failed to get image auth for https://index.docker.io/v1/. Setting empty credentials for the image: cloud-spanner-emulator/emulator:latest. Error is:credentials not found in native keychain
2023/05/22 01:57:23 Failed to get image auth for https://index.docker.io/v1/. Setting empty credentials for the image: cloud-spanner-emulator/emulator:latest. Error is:credentials not found in native keychain
    read_batch_test.go:46: failed to start container with 5 retries: Error response from daemon: pull access denied for cloud-spanner-emulator/emulator, repository does not exist or may require 'docker login': denied: requested access to the resource is denied: failed to create container
--- FAIL: TestSpannerIO_QueryBatch (29.49s)
FAIL
FAIL	github.com/apache/beam/sdks/v2/go/test/integration/io/spannerio	29.527s
```

lostluck commented:
Run Go PostCommit

lostluck commented:
Ah looks like the job that failed on the last run https://ci-beam.apache.org/job/beam_PostCommit_Go_PR/714/consoleText actually passed, and the dataflow API call just didn't succeed. Merging.
